book/layout.md

```diff
@@ -232,54 +225,66 @@ BLOCK_ELEMENTS = [
 ]
 ```
 
-We'll use `BlockLayout` for elements with children in that list, and
-`InlineLayout` otherwise. Put that logic in a new `layout_mode`
-function:
+Our `layout_mode` function has to handle one tricky case, where a node
+contains both block objects like a `<p>` element but also text objects
+like a text node or a `<b>` element. I've chosen to use block mode in
+this case, but it's probably best to think of this as a kind of error
+on the part of the web developer. And just like with implicit tags in
+[Chapter 4](html.md), we use a repair mechanism to make sense of the
+situation.[^anon-block]
+
+[^anon-block]: In real browsers, that repair mechanism is called
+"[anonymous block boxes][anon-block]" and is more complex than what's
+described here.
+
+So now `BlockLayout` can determine what kind of layout to do based on
+the `layout_mode` of its HTML node:
 
 ``` {.python}
-def layout_mode(node):
-    if isinstance(node, Text):
-        return "inline"
-    elif node.children:
-        for child in node.children:
-            if isinstance(child, Text): continue
-            if child.tag in BLOCK_ELEMENTS:
-                return "block"
-        return "inline"
-    else:
-        return "block"
+class BlockLayout:
+    def layout(self):
+        mode = layout_mode(self.node)
+        if mode == "block":
+            previous = None
+            for child in self.node.children:
+                # ...
+        else:
+            # ...
+            self.recurse(self.node)
+            self.flush()
 ```
 
-This function additionally makes sure text nodes get inline layout
-while empty elements get block layout. Now we can call `layout_mode`
-to determine which layout mode to use for each element:
+Finally, since `BlockLayout`s can now have children, the `layout`
+method needs to recursively call `layout` so those children can
+construct their children, and so on recursively:
 
 ``` {.python}
 class BlockLayout:
     def layout(self):
-        previous = None
-        for child in self.node.children:
-            if layout_mode(child) == "inline":
-                next = InlineLayout(child, self, previous)
-            else:
-                next = BlockLayout(child, self, previous)
-            self.children.append(next)
-            previous = next
         # ...
+        for child in self.children:
+            child.layout()
 ```
 
-Our layout tree now has a `DocumentLayout` at the root, `BlockLayout`s
-at interior nodes, and `InlineLayout`s at the leaves:[^or-empty]
+We also need to gather their `display_list` fields into a single
+array:
 
-[^or-empty]: Or, the leaf nodes could be `BlockLayout`s for empty
-elements.
+``` {.python expected=False}
+class BlockLayout:
+    def layout(self):
+        # ...
+        for child in self.children:
+            self.display_list.extend(child.display_list)
+```
 
-::: {.widget big-height=490px small-height=860px}
```

Comment:
Don't want this example any more?

Reply:
No, you're right, this one is still good (but I need to edit the caption). The other one (which had just `BlockLayout`s) is no longer good, since now the first stage is only inline, not only block.